JWT for User Auth in API Calls

MrAaronOlsen opened the pull request:
# API calls from React are now authorized via JWT tokens.
- JWT token is created when a user visits a page
- The token is actually the entire user model tokenized. This is sent to React as the encrypted token.
- NOTE: All I really need is the id since I look for the User again on the API side. However, I thought it might be fine to simple check if 'role' == 'admin' from the user token. Not sure what is best practice here.
- Token is sent in the header of a request using "authorization Bearer"
- It can be found at request.env["HTTP_AUTHORIZATION"]
# Reasoning
- Having an API leaves open the possibility of someone being able to access that data. Without protections anyone could call the API routes and get back sensitive data.
- Even from an internal perspective there is a need to scope resources to users. With JWT tokens we can easily and securely do this.
- This gives some flexibility down the line. Aside from initial OAuth and portal entry, the frontend is more or less separated from the back.

@srt32

MrAaronOlsen commented:
This is an important feature and I need to make sure all the bases are being covered.

srt32 commented on app/controllers/admin/dashboard_controller.rb:
```diff
@@ -2,6 +2,7 @@ class Admin::DashboardController < Admin::BaseController
 
   def index
     @cohorts = Cohort.all.to_json(:include => [:applications, :reviewers])
+    @user = jwt_encode({user: current_user.to_json})
```
I would recommend encoding just the `id` and then looking it up on the server. It reduces the amount of data we're passing around and keeps things simpler until we need them to be more comlicated.

MrAaronOlsen replied:
I agree. Easy fix.

MrAaronOlsen replied:
Token is now just the User ID.

srt32 commented on app/controllers/api_controller.rb:
```diff
@@ -0,0 +1,18 @@
+class ApiController < ActionController::API
```
Please add some specs around this work. We'd want to make sure that

* encoding and decoding works in the happy path.
* probably want to set an expiry on the token and make sure it expires properly

MrAaronOlsen replied:
I'm working on getting testing set up for React and will spend a good chunk of time getting myself up to date here.

MrAaronOlsen replied:
Yes, I'm adding the backend Unit specs as well. Will be doing this next.

srt32 commented on app/controllers/api_controller.rb:
```diff
@@ -0,0 +1,18 @@
+      end
+
+      def jwt_decode(token)
+        decoded = JWT.decode(token, Rails.application.secrets.secret_key_base)
```
What's the behavior here if the token is invalid or expired? Does it raise? if it does should we handle it?

MrAaronOlsen replied:
Yes, I don't have error checks yet. I'll add those. Right now we just get a Rails error.

MrAaronOlsen replied:
Error rescuing in place. Simple for now, but will catch all JWT decode errors.

srt32 commented on app/controllers/application_controller.rb:
```diff
@@ -9,4 +9,8 @@ class ApplicationController < ActionController::Base
     end
+
+    def jwt_encode(payload)
+      JWT.encode(payload, Rails.application.secrets.secret_key_base)
```
Why do we encode here if all the other jwt logic is in api_controller?

MrAaronOlsen replied:
We encode here because this is where a user enters the app. In most situations a user would call the API and get a token back, but in this case it's not really necessary since they're already authenticated. I just pass this token on to React as a prop when they're routed to a dashboard.

srt32 commented:
Looks good overall. I would advocate for a bunch of unit specs around the token usages.

srt32 commented:
I would add a bunch of backend unit specs too
On Thu, Sep 14, 2017 at 08:28 Aaron Olsen <notifications@github.com> wrote:

> *@MrAaronOlsen* commented on this pull request.
> ------------------------------
>
> In app/controllers/api_controller.rb
> <https://github.com/turingschool/full-circle/pull/41#discussion_r138927794>
> :
>
> > @@ -0,0 +1,18 @@
> +class ApiController < ActionController::API
>
> I'm working on getting testing set up for React and will spend a good
> chunk of time getting myself up to date here.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/turingschool/full-circle/pull/41#discussion_r138927794>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACFI7GBbE-n1iQPurspbdeobA7JC11Fnks5siUYngaJpZM4PW_Xz>
> .
>
-- 
Support our ride in the PMC to fight cancer! http://profile.pmc.org/ST0098

MrAaronOlsen commented:
@srt32 I addressed most of the issues you raised here. I also included some of the small fixes you requested on Tuesday's meeting including https and unique ids on uid. Readme is also updated.
